src/realm/query_engine.hpp

```diff
@@ -2508,86 +2515,66 @@ class TwoColumnsNode : public ParentNode {
+    TwoColumnsNodeBase(const TwoColumnsNodeBase& from)
+        : ParentNode(from)
+        , m_condition_column_key1(from.m_condition_column_key1)
+        , m_condition_column_key2(from.m_condition_column_key2)
     {
-        size_t s = start;
+    }
 
-        while (s < end) {
-            if (std::is_same<TConditionValue, int64_t>::value) {
-                // For int64_t we've created an array intrinsics named compare_leafs which template expands bitwidths
-                // of boths arrays to make Get faster.
-                QueryState<int64_t> qs(act_ReturnFirst);
-                bool resume = m_leaf_ptr1->template compare_leafs<TConditionFunction, act_ReturnFirst>(
-                    m_leaf_ptr2, start, end, 0, &qs, CallbackDummy());
-
-                if (resume)
-                    s = end;
-                else
-                    return to_size_t(qs.m_state);
-            }
-            else {
-// This is for float and double.
+protected:
+    mutable ColKey m_condition_column_key1;
+    mutable ColKey m_condition_column_key2;
+    std::unique_ptr<ArrayPayload> m_leaf_ptr1 = nullptr;
+    std::unique_ptr<ArrayPayload> m_leaf_ptr2 = nullptr;
+};
 
-#if 0 && defined(REALM_COMPILER_AVX)
-// AVX has been disabled because of array alignment (see https://app.asana.com/0/search/8836174089724/5763107052506)
-//
-// For AVX you can call things like if (sseavx<1>()) to test for AVX, and then utilize _mm256_movemask_ps (VC)
-// or movemask_cmp_ps (gcc/clang)
-//
-// See https://github.com/rrrlasse/realm/tree/AVX for an example of utilizing AVX for a two-column search which has
-// been benchmarked to: floats: 288 ms vs 552 by using AVX compared to 2-level-unrolled FPU loop. doubles: 415 ms vs
-// 475 (more bandwidth bound). Tests against SSE have not been performed; AVX may not pay off. Please benchmark
-#endif
 
-                TConditionValue v1 = m_leaf_ptr1->get(s);
-                TConditionValue v2 = m_leaf_ptr2->get(s);
-                TConditionFunction C;
+template <class TConditionFunction>
+class TwoColumnsNode : public TwoColumnsNodeBase {
+public:
+    using TwoColumnsNodeBase::TwoColumnsNodeBase;
+    ~TwoColumnsNode() noexcept override {}
+    size_t find_first_local(size_t start, size_t end) override
+    {
+        size_t s = start;
+        ColumnType col1_type = m_condition_column_key1.get_type();
+        ColumnType col2_type = m_condition_column_key2.get_type();
+        bool col1_nullable = m_condition_column_key1.is_nullable();
+        bool col2_nullable = m_condition_column_key2.is_nullable();
 
-                if (C(v1, v2))
-                    return s;
-                else
-                    s++;
-            }
+        while (s < end) {
+            Mixed v1 = get_value_from_leaf(m_leaf_ptr1.get(), col1_type, col1_nullable, s);
+            Mixed v2 = get_value_from_leaf(m_leaf_ptr2.get(), col2_type, col2_nullable, s);
+            if (TConditionFunction()(v1, v2))
```

Comment:
how about v1.is_null() and v2.is_null()?

Reply:
The Mixed overloaded operator methods ignore the null parameters because they have to check nullability as part of the type compatibility check. I guess I was hoping to not add them as an optimization, but I can add them in to future proof these calls and for consistency. Since they are unused, I think the compiler might be able to optimize them away anyways.

Reply:
Ok, you're right